left_sidebar: Add "Alert words" view with unread count.

santhoshh-kumar opened the pull request:
<!-- Describe your pull request here.-->

This PR introduces alert words view to the left sidebar as described in the issue description. Most of the changes in this PR follows the same pattern used for mentions. 

A small thing to notice is that we are not live updating the unread counter of "Alert words" in the `...` menu popover when the views section collapsed because it's just a popover which the user will open temporarily and then close. 

The icon used is temporary since it was mentioned as `TBD` in the issue description.

One more thing to notice is that even if the user has some unread messages in the "Alert Words" view, it will get immediately hidden from the left sidebar if the user removes all the configured alert words. I think it makes sense to hide it but still bringing it to your notice.

As a follow up to this PR, we can also consider adding alert words icon badge in the recent view UI similar to how we are doing for mentions. 

This PR doesn't include doc updates. We need to do doc updates after this PR.

Fixes: #28778

**How changes were tested:**

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

I've included some screen captures and screenshots to show how the changes were tested. 

**Screenshots and screen captures:**

<details>
<summary>On message send</summary>

![message-send](https://github.com/user-attachments/assets/a91bfd24-9636-4645-802a-e9a3a67f97df)

</details>

<details>
<summary>On message delete</summary>

![message-delete](https://github.com/user-attachments/assets/6656bbab-7b6a-4fc3-a300-74cc313f107b)

</details>

<details>
<summary>On message edit</summary>

![message-edit](https://github.com/user-attachments/assets/e292d713-8c66-4f94-911b-6f769b36dd37)

</details>

<details>
<summary>On read</summary>

![message-read](https://github.com/user-attachments/assets/8b36af8c-9947-4f00-9945-7c64f21a5513)

</details>

<details>
<summary>Live update show/hide alert words view</summary>

![alert-words](https://github.com/user-attachments/assets/6d9a4a3a-a6e5-43b6-a963-a714c050c9fc)


</details>


<details>
<summary>UI screenshots</summary>

<img width="747" height="491" alt="image" src="https://github.com/user-attachments/assets/bd065db9-35ff-4eb0-a8ce-aeb2e2376c3b" />  

---

<img width="747" height="491" alt="image" src="https://github.com/user-attachments/assets/f5b3e137-956b-4b04-b815-7fa801cf04b7" />

---

<img width="984" height="476" alt="image" src="https://github.com/user-attachments/assets/ba5c6d63-9b17-4821-ae4f-67cb8e1999e6" />



</details>

---
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

santhoshh-kumar committed:
watched_phrases: Add support to unread messages.

This change adds support for tracking and returning watched phrases (alert words)
in unread message data. When a user has configured watched phrases (alert words),
unread messages containing those words will now be flagged appropriately in the API
responses for unread messages. This follows the existing pattern for mentions.

This backend change allows us to live update the counter for unread messages
on message send in alert words view after introducing it to the left_sidebar
similar to mentions view.

Fixes part of #28778.

santhoshh-kumar committed:
watched_phrases: Add support to message edits.

On message edit, there might be a possibility for watched phrases (alert words)
to be added or removed from the message. This change helps track them by following
the existing pattern for mentions.

This backend change allows us to live update the counter for unread messages
on message edit in alert words view after introducing it to the left_sidebar
similar to mentions view. It also allows to send notification if the message
contains alert words after edit.

Fixes part of #28778.

santhoshh-kumar committed:
unread: Add alert word tracking to unread message counts.

This adds support for tracking unread messages that contain alert
words, similar to how we track unread messages with mentions.

This change is the foundation to track the counter for unread
messages in alert words view after introducing it to the left_sidebar.

santhoshh-kumar committed:
views: Add alert words view to navigation system.

Adds 'Alert words' as a built-in navigation view with proper filter support,
icon, and empty state banner.

This change will allow to include alert words in the left sidebar navigation
area since get_built_in_views() function returns all the built-in views
including alert_words.

santhoshh-kumar committed:
left sidebar: Add alert words row with unread counts.

Adds a dedicated row in the left sidebar navigation area for alert words,
displaying unread counts and applying animations for new unreads.

This row is only shown when the user has alert words configured. The unread
counter will get live updated on message receive, edit and delete.

Fixes #28778.

santhoshh-kumar committed:
server_events: Update alert words row on alert_words event.

When the user updates their alert words, the left sidebar navigation
area should update to show or hide the alert words row based on
whether the user has any alert words configured.

Since we receive an alert_words event when the user updates their
alert words, this commit updates the row when we receive this event.

santhoshh-kumar committed:
tooltip: Add tooltip for alert words view.

Adds a tooltip for the alert words view in the left sidebar similar
to mentions indicating number of unread alert word messages.

santhoshh-kumar committed:
popover_menus: Add popover menu for alert words view.

Adds a popover menu for the alert words view in the left sidebar
to easily navigate to the alert words settings for configuring
alert words.

alya commented:
Thanks for the detailed screen captures!

roanster007 commented:
Sure, I'll get it done in 1-2 days

roanster007 commented on web/src/left_sidebar_navigation_area.ts:
```diff
@@ -290,6 +318,13 @@ export function get_built_in_popover_condensed_views(): navigation_views.BuiltIn
+            // Hide alert words view if user has no alert words configured
+            if (!alert_words.has_alert_words_configured()) {
+                return false;
+            }
```
We probably don't want to hide the view if no alert word is present? Probably better to have a #design discussion about this on CZO

santhoshh-kumar replied:
Actually this is stated in the issue description already. So I don't think we need to discuss about it on CZO. But there was no discussion about it already either. What do you think?

roanster007 commented on zerver/tornado/event_queue.py:
```diff
@@ -1541,9 +1544,9 @@ def maybe_enqueue_notifications_for_message_update(
-        # especially important considering that most message
+    if prior_mentioned or prior_alerted:
+        # Don't spam people with duplicate mentions or alert words.
+        # This is especially important considering that most message
```
I don't quite recall the significance of this part, do you mind giving a short summary of what's happening here

santhoshh-kumar replied:
If  we have notified the user for the message containing alert word / mentions already, we don't need to notify again after editing the message because a lot of times it can be a simple typo correction. If we do notify for every edits - it will lead to spamming the user with duplicate mentions or alert words.

santhoshh-kumar commented on web/src/left_sidebar_navigation_area.ts:
```diff
@@ -173,11 +198,14 @@ function toggle_condensed_navigation_area(): void {
+    new_count: number,
+    previous_count: number,
+): void {
+    if (new_count > previous_count) {
```
We are using the same function for animating both `alert_words` and `mentions`. So we can't use it like that.

santhoshh-kumar commented:
Thanks for your review @roanster007 
I've seperated the `unread_alert_word` test from `unread_mention` tests.

roanster007 reviewed (approved):
LGTM!

santhoshh-kumar committed:
unread_ops: Add alert word tracking on marking as unread.

This adds support for tracking messages that contain alert words on
marking as unread.

This change is the foundation to track the counter when marking the
messages as unread after introducing it to the left_sidebar.

santhoshh-kumar commented:
> you mark some messages as unread, but the unread counter doesn't show up

Oops, I didn't see that we could mark messages as unread!

> How hard is this to fix?

Looking at the changes, I've found out that I have updated the backend `MessageDetailsDict` but not have not made frontend changes for it. I've fixed it by adding [a commit](https://github.com/zulip/zulip/pull/36764/commits/eabf062dbd9085c3905df89e1e663772c0859789). 

<details>
<summary>Screen Capture</summary>

![mark-as-unread](https://github.com/user-attachments/assets/e0324c0f-9378-4fd0-ac9a-0dfdcd24d81e)

</details>

santhoshh-kumar commented:
In the above screen capture there is no animation when the counter updates the second time for `alert words`.

When I looked into this issue deeply, I found out that if the counter updates while the previous animation is in its final few milliseconds, the new update does not trigger the animation.

Then I checked whether this issue present in the `main` for `mentions` unread counter animation. And yes, this issue is pre-existing.

It's not a major issue, but I just wanted to bring it to your attention.

santhoshh-kumar commented:
@zulipbot remove "has conflicts"

whilstsomebody commented:
@santhoshh-kumar , are you working on this? If not, I would like to pick this up to push to completion.

santhoshh-kumar commented:
@whilstsomebody - Yeah, you can push this to the completion. Thanks!
